Fix `fmt::format` crash in raft topology code

kbr-scylla opened the pull request:
Fix a crash on `aarch64` in release mode on `fmt::format` call; replacing it with `seastar::format` helps.

Fixes #13707

tchaikov commented on service/storage_service.cc:
```diff
@@ -436,7 +436,7 @@ future<> storage_service::topology_transition(storage_proxy& proxy, cdc::generat
     } catch (replica::no_such_column_family& e) {
         slogger.error("Error while applying topology mutations from {}: {}", from, e);
-        throw std::runtime_error(fmt::format("Error while applying topology mutations: {}", e));
+        throw std::runtime_error(format("Error while applying topology mutations: {}", e));
```
this reminds me that @avikivity also fixed a similar issue in Seastar.  see https://github.com/scylladb/seastar/pull/1249 . that issue only exhibits itself on aarch64.

tchaikov replied:
i need to take a closer look at this backtrace pasted by Avi.

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -471,7 +471,7 @@ class topology_mutation_builder {
     template<typename T>
     topology_mutation_builder& set(const char* cell, const T& value) {
-        return set(cell, sstring{fmt::format("{}", value)});
+        return set(cell, sstring{format("{}", value)});
```
`format` already returns `sstring` so the cast can be dropped.

kbr-scylla commented:
v2:
- rebased
- use `format(...)` instead of `sstring{format(...)}`

avikivity commented on service/storage_service.cc:
```diff
@@ -434,7 +434,7 @@ future<> storage_service::topology_transition(storage_proxy& proxy, cdc::generat
     } catch (replica::no_such_column_family& e) {
         slogger.error("Error while applying topology mutations from {}: {}", from, e);
-        throw std::runtime_error(fmt::format("Error while applying topology mutations: {}", e));
+        throw std::runtime_error(format("Error while applying topology mutations: {}", e));
```
Please use seastar::format or ::format (same thing). Unqualified format will also match std::format using argument-dependent format, and so will be ambiguous, once we switch to libstdc++ 13.

avikivity replied:
below as well.

nyh commented on service/migration_manager.cc:
```diff
@@ -902,7 +902,7 @@ future<> migration_manager::push_schema_mutation(const gms::inet_address& endpoi
 }
 
-future<> migration_manager::announce_with_raft(std::vector<mutation> schema, group0_guard guard, std::string_view description) {
+future<> migration_manager::announce_with_raft(std::vector<mutation> schema, group0_guard guard, const sstring& description) {
```
I think you've opened a huge pandora's box here.
First, there is a significant risk that nobody will remember in the future why you you changed to const sstring& here, and will change it back to std::string_view and nobody will notice this.
Second, this problem isn't at all specific to this specific functions - in many many functions all across the code, we've been slowly changing "const sstring&" to "std::string_view". All of those may be accidentally used with a temporary and risk this bug. I'm worried that maybe we have a dozen other instances of this bug and don't know...
I see this is a known pitfall - https://www.foonathan.net/2017/03/string_view-temporary/ - and an attempt to solve this http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0936r0.pdf.

kbr-scylla replied:
I would say changing to `string_view` is what's opening the pandora's box.

I can leave a comment explaining why I use `const sstring&`.

kbr-scylla replied:
Hmm but wait a sec...

the article you sent made me realize - calling `f(format(...))` where `f` takes a `string_view` should actually be safe, because the lifetime of `format(...)` ends not at the moment of entering `f`, but at the end of the entire expression `f(format(...))`!

However does it also apply for expressions such as `co_await f(format(...))`? Will the `format(...)` be destroyed only after the `co_await` is evaluated, or can it be destroyed as soon as `f` returns?

kbr-scylla replied:
> However does it also apply for expressions such as co_await f(format(...))? Will the format(...) be destroyed only after the co_await is evaluated, or can it be destroyed as soon as f returns?

Well, `f(format(...))` is not a full expression inside `co_await f(format(...))` according to the definition of "full expression", so yes, the temporary should be kept alive.
See also:
https://stackoverflow.com/questions/45744673/c-full-expression-standard-wording
https://stackoverflow.com/questions/62620147/c-coroutines-temporaries-in-co-await-expressions

I'll remove this first commit then.

nyh replied:
> Hmm but wait a sec...
> 
> the article you sent made me realize - calling `f(format(...))` where `f` takes a `string_view` should actually be safe, because the lifetime of `format(...)` ends not at the moment of entering `f`, but at the end of the entire expression `f(format(...))`!

Right, I was also reading this article now and realized the same thing...

> However does it also apply for expressions such as `co_await f(format(...))`? Will the `format(...)` be destroyed only after the `co_await` is evaluated, or can it be destroyed as soon as `f` returns?

Right! The problem is not the co_await but the asynchronous function itself - migration_manager::announce_with_raft -  which returns a future, and takes a string_view.
It can't "save" this string_view after "returning" a future (i.e., after the first preemption point).

So maybe that should be our general rule (and documented somewhere?) - string_view are great in normal, synchronous functions, but should be avoided asynchronous functions.
Or not avoided - but used with care. same as references where we have "conventions" about this (passing a reference to an asynchronous function usually means you promise to hold this reference alive until the future is complete).

kbr-scylla replied:
>  It can't "save" this string_view after "returning" a future (i.e., after the first preemption point).

Apparently it's "saved" until the full `co_await` expression ends, so it should be kept alive (according to the standard - compilers sometimes do their own thing...) even across yields.

tgrabiec replied:
> passing a reference to an asynchronous function usually means you promise to hold this reference alive until the future is complete

Same rule applies to views. Accepting `const sstring&` instead of a view won't protect you from a user which passes a temporary sstring object.

gleb-cloudius commented:
On Thu, May 04, 2023 at 03:29:22AM -0700, nyh wrote:
> I see this is a known pitfall - https://www.foonathan.net/2017/03/string_view-temporary/

This says that the way it was used here is correct actually.

--
			Gleb.

kbr-scylla commented:
v3:
- removed the first commit, turns out it was unnecessary
- use `::format` instead of `format`; I replaced all uses of `format` with `::format` in service/storage_service.cc for consistency.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1045/

kbr-scylla commented:
or @nyh

yaronkaikov commented:
@avikivity @tgrabiec Can you please review and merge?

tgrabiec commented on service/storage_service.cc:
```diff
@@ -377,7 +377,7 @@ future<> storage_service::topology_state_load(cdc::generation_service& cdc_gen_s
                 } else {
                     tmptr->add_bootstrap_tokens(rs.ring.value().tokens, ip);
-                    co_await update_pending_ranges(tmptr, format("bootstrapping node {}/{}", id, ip));
+                    co_await update_pending_ranges(tmptr, ::format("bootstrapping node {}/{}", id, ip));
```
Maybe the problem is actually that the formatter for tagged_uuid is defined like this:

```
struct fmt::formatter<utils::tagged_uuid<Tag>> : fmt::formatter<std::string_view>
```

Notice that it extends a formatter for string_view. Maybe some fmt impl has a specialization for this formatter, and misinterprets the value as string view? Can you check if avoiding the inheritance would fix the problem?

tgrabiec replied:
The fact that we see this invoked in the backtrace would support this:
```
fmt::v9::vformat[abi:cxx11](fmt::v9::basic_string_view<char>, fmt::v9::basic_format_args<fmt::v9::basic_format_context<fmt::v9::appender, char> >) at ??:?
```

kbr-scylla replied:
Well, it doesn't compile without the inheritance
```
/usr/include/fmt/core.h:1311:28: error: no member named 'parse' in 'fmt::formatter<utils::UUID>'                                                                                                                                                                                       
    parse_ctx.advance_to(f.parse(parse_ctx));                                                                                              
                         ~ ^                                                                                                               
/usr/include/fmt/core.h:1293:21: note: in instantiation of function template specialization 'fmt::detail::value<fmt::basic_format_context<fmt::appender, char>>::format_custom_arg<utils::UUID, fmt::formatter<utils::UUID>>' requested here
    custom.format = format_custom_arg<                                                                                                     
                    ^                                                                                                                      
/home/kbraun/scylladb/seastar/include/seastar/util/log.hh:226:33: note: in instantiation of function template specialization 'fmt::format_to<seastar::internal::log_buf::inserter_iterator, const utils::UUID &, unsigned long, const unsigned int &, const unsigned int &, 0>' request
ed here                                                                                                                                    
                    return fmt::format_to(it, fmt::runtime(fmt.format), std::forward<Args>(args)...);                                      
                                ^                                                                                                                                                                                                                                                      
/home/kbraun/scylladb/seastar/include/seastar/util/log.hh:385:9: note: in instantiation of function template specialization 'seastar::logger::log<const utils::UUID &, unsigned long, const unsigned int &, const unsigned int &>' requested here
        log(log_level::trace, std::move(fmt), std::forward<Args>(args)...);                                                                
        ^                                                                                                                                  
./tracing/trace_state.hh:136:28: note: in instantiation of function template specialization 'seastar::logger::trace<const utils::UUID &, unsigned long, const unsigned int &, const unsigned int &>' requested here
        trace_state_logger.trace("{}: props {}, slow query threshold {}us, slow query ttl {}s", session_id(), _state_props.mask(), info.slow_query_threshold_us, info.slow_query_ttl_sec);
```

also note that this is private inheritance.

avikivity replied:
You could define parse() instead. But I don't know if inheriting from string_view's formatter is good or not, I saw it in some online examples.

kbr-scylla replied:
I don't see how "misinterpreting the value as string_view" could happen with private inheritance though.

AFAIU it's only used so we can implement `format` conveniently, delegating to `formatter<string_view>::format`.

tgrabiec replied:
Note that it's a struct, so it's actually a public inheritance.

kbr-scylla replied:
> Note that it's a struct, so it's actually a public inheritance.

Oh.
